nengo/builder/optimizer.py

```diff
@@ -11,386 +10,707 @@
+        else:
+            initial_value = signals[0].initial_value
+        readonly = all(s.readonly for s in signals)
+        name = 'merged<' + signals[0].name + ', ..., ' + signals[-1].name + '>'
+        merged_signal = Signal(initial_value, name=name, readonly=readonly)
+
+        if signals[0].ndim > 0:
+            replacements = {}
+            start = 0
+            for s in signals:
+                size = s.shape[axis]
+                indexing = [slice(None)] * initial_value.ndim
+                indexing[axis] = slice(start, start + size)
+                replacements[s] = merged_signal[tuple(indexing)]
+                start += size
+        else:
+            replacements = {s: merged_signal for s in signals}
+
+        return merged_signal, replacements
+
+    @staticmethod
+    def merge_views(signals, axis=0):
+        """Merges multiple signal views into one continguous signal view.
+
+        Parameters
+        ----------
+        signals : sequence
+            Signals to merge. Must only contain views.
+        axis : int, optional
+            Axis along which to concatenate the signals.
+
+        Returns
+        -------
+        merged_signal : Signal
+            The merged signal.
+        replacements : dict
+            Dictionary mapping from the old signals to new  signals that are
+            a view into the merged signal. Used to replace old signals.
+        """
+        _SigMerger.check_views(signals, axis=axis)
+
+        shape = (
+            signals[0].shape[:axis] + (sum(s.shape[axis] for s in signals),) +
+            signals[0].shape[axis+1:])
+        initial_value = np.ndarray(
+            buffer=signals[0].base.initial_value, dtype=signals[0].dtype,
+            shape=shape, offset=signals[0].offset, strides=signals[0].strides)
+        merged_signal = Signal(
+            initial_value, name=signals[0].base.name, base=signals[0].base,
+            readonly=all(s.readonly for s in signals))
+
+        return merged_signal, {}  # TODO: why no replacements here?
+
+
+class _OpMerger(object):
+    op_checkers = {}
+    sig_checkers = {}
+    op_mergers = {}
 
     @classmethod
-    def can_merge_ops(cls, op1, op2):
+    def is_mergeable(cls, op, tomerge):
         return (
-            type(op1) == type(op2) and
-            type(op1) in cls.op_mergers and
-            cls.op_mergers[type(op1)].can_merge(op1, op2))
+            tomerge.optype in cls.op_checkers and
+            type(op) is tomerge.optype and
+            op not in tomerge.merged and
+            op not in tomerge.all_tc and
+            len(tomerge.tc[op].intersection(tomerge.ops)) == 0 and
+            all(tomerge.check_signals(o, op) for o in tomerge.ops) and
+            all(op not in tomerge.tc[o] and o not in tomerge.tc[op]
+                for o in tomerge.merged) and
+            cls.sig_checkers[tomerge.optype](op, tomerge) and
+            cls.op_checkers[tomerge.optype](tomerge.last_op, op))
```

Comment:
I find this hard to parse because it is a long list of conditions without their intend. Prior to the merge I assigned groups of these conditions to variables as a way of documenting, for example `indepent_of_prior_merges` -> “these are the conditions for this operator to be independent of prior merges”. This gives then also a much more human readable statement in the end when combining those variables.